src/orion/core/io/config.py

```diff
@@ -0,0 +1,278 @@
+
+
+logger = logging.getLogger(__name__)
+
+
+NOT_SET = object()
+
+
+class ConfigurationError(Exception):
+    """Error raised when a configuration value is requested but not set."""
+
+
+class Configuration:
+    """Configuration object
+
+    Provides default values configurable at different levels. The configuration object can have
+    global default values, which may be overriden by user with yaml configuration files, environment
+    variables or by setting directly the values in the configuration object. In order, direct
+    definition overrides, environment variables, which overrides yaml configuration, which overrides
+    default values in configuration object definition.
+
+    Examples
+    --------
+    >>> config = Configuration()
+    >>> config.add_option('test', str, 'hello', 'TEST_ENV')
+    >>> config.test
+    'hello'
+    >>> config.load_yaml('some_config.yaml')
+    >>> config.test
+    'in yaml!'
+    >>> os.environ['TEST_ENV'] = 'there!'
+    >>> config.test
+    'there!'
+    >>> config.test = 'here'
+    >>> config.test
+    'here'
+
+    """
+
+    def __init__(self):
+        self.config = {}
+
+    def load_yaml(self, path):
+        """Load yaml file and set global default configuration
+
+        Parameters
+        ----------
+        path: str
+            Path to the global configuration file.
+
+        Raises
+        -------
+        ConfigurationError
+            If some options in the yaml file does not exist in the config
+
+        """
+        with open(path) as f:
+            cfg = yaml.safe_load(f)
+            if cfg is None:
+                return
+            # implies that yaml must be in dict form
+            for key, value in flatten(cfg).items():
+                default = self[key]
+                logger.debug('Overwritting "%s" default %s with %s', key, default, value)
+                self.config[key]['yaml'] = value
+
+    def __getattr__(self, key):
+        """Get the value of the option
+
+        Parameters
+        ----------
+        key: str
+            Name of the option
+
+        Returns
+        -------
+            Value of the option.
+
+        Raises
+        -------
+        ConfigurationError
+            If the option does not exist
+
+        """
+        if key == 'config':
+            raise AttributeError
+        if key not in self.config:
+            raise ConfigurationError("Configuration does not have an attribute "
+                                     "'{}'.".format(key))
+
+        config_setting = self.config[key]
+        if 'value' in config_setting:
+            value = config_setting['value']
+        elif ('env_var' in config_setting and
+              config_setting['env_var'] in os.environ):
+            value = os.environ[config_setting['env_var']]
+        elif 'yaml' in config_setting:
+            value = config_setting['yaml']
+        elif 'default' in config_setting:
+            value = config_setting['default']
+        else:
```

Comment:
I think we could do some kind of more good-looking loop on this, no?

Reply:
Not sure, this really is a `switch` case.